src/IO/S3/URI.cpp

```diff
@@ -154,6 +155,24 @@ URI::URI(const std::string & uri_, bool allow_archive_path_syntax)
             key = uri.getPath().substr(1);
     }
 
+    /// If there is a '?' in the original string, but no actual query, it means
+    /// the user intended to use '?' as a wildcard in the path
+    if (original_uri.getRawQuery().empty() && uri_str.find('?') != std::string::npos && !has_version_id && !looks_like_presigned)
```

Comment:
How are urls with trailing '?' should be treated? It think it worths to add a unit test for that.